### DISCUSSION TITLE
User extensions

### DISCUSSION THREAD

**Discussion by tavin** [executablebooks/meta#962] [Ideas] [answered]
I have a couple of ideas/wishes related to markdown parsing and rendering, and it seems a MyST syntax extension might be the natural (only?) way to realize them. However the barrier to entry is perhaps rather high. Do I submit a MEP? Do I need to make contacts in multiple organizations for that?

I can think of a couple ways to make it much easier for a student/hobbyist to get started on something like this:
* Processing of `enable_extensions` could treat a dotted string as a reference to an external object which satisfies an extension interface.
* There could be a separate `user_extensions` list with like behavior.

From what I've seen I suppose the interface would just be a callable that receives the markdownit parser.

**Comment by chrisjsewell** [comment, marked as answer]
See my answer in https://github.com/executablebooks/MyST-Parser/issues/632#issuecomment-1455109136

**Reply by tavin** [reply]
Hi @chrisjsewell ok that's fair. I was just looking for a nice way to begin developing an idea until it's baked enough to try proposing changes in the spec, MEPs, etc. And with a straightforward way for others to view the outcome, if they have a copy of your "extension" code.

If maintaining a fork/patch is the only reasonable way so be it.

**Reply by chrisjsewell** [reply]
Yeh I think a fork is best for prototyping, and obviously happy to hear your ideas and provide any feedback

**Reply by agoose77** [reply]
@tavin to add more context here; as you and @chrisjsewell note, the "canonical" way to add new syntax is to do so at the MEP level, at which point it becomes an official part of MyST. 

In _general_, though, we haven't formalised any solutions around directives/roles extensions that do _not_ require MEPs. For example, if you want to add a custom role, it should be possible for all parsers to understand the grammar of a role/directive. I am fairly confident that this is already the case; myst-js does this, I *think* our markdown-it stack does this, and I presume that myst-parser does too, though the maintainers there (ping @chrisjsewell) can answer that directly.

Adding custom roles and directives is conceivably within the scope of a domain-specific MyST distribution. The real challenge, of course, is what to do with the parsed content, and that's what I'm really referring to when I say that there's no formal solution. I don't think any of our renderers right now offer a mechanism for adding custom directive renderers, though it should be possible for all the stacks that I can think of.

To summarise, there are implications for allowing arbitrary roles and directives, and although I suspect we'd vote in favour of making that easier, it opens a can of worms that we'd collectively need to address!

**Reply by chrisjsewell** [reply]
> In general, though, we haven't formalised any solutions around directives/roles extensions

@agoose77 I think you have misunderstood what we are discussing here? 
its not to do about directives/roles extensions, its to do with  adding https://github.com/executablebooks/mdit-py-plugins syntax parsing extensions to myst-parser

myst-parser, via sphinx, already offers a rich, and fully though out, extension mechanism that is much more than just roles and directives: https://www.sphinx-doc.org/en/master/extdev/index.html#dev-extensions
(sphinx has its problems, but the extension mechanism is pretty good)

myst-js role/directive mechanism, which is mainly copied from the code I wrote in myst-docutils (and was really just a hack to start doing vs code previews), essentially involves doing low level markdown-it specific parsing.
I was trying to propose moving past this in https://github.com/executablebooks/unified-myst/tree/main/packages/core-parse#extension-mechanism,
unfortunately no one took me up on the suggestion 🤷

**Reply by agoose77** [reply]
@chrisjsewell no, I'm talking to the broader point because I suspect this issue will be found for related queries (we're in EB/discussions here!). I wanted to reference the fact that beyond the MEP process we don't have a cross-platform solution for changing syntax, and that meanwhile there is a limited subset (new directives and roles) that we could build-out more if we align on it as a priority.

From @tavin's post, it seems like the main focus is syntax, but given the scope of the issue, the rest is for posterity :)

Though, for other readers: `enable_extensions` + MarkdownIt infers that this is a specific question about `myst-parser`.

**Reply by chrisjsewell** [reply]
> no, I'm talking to the broader point because I suspect this issue will be found for related queries

I'm happy to discuss user extensions more broadly, but then it would be best to open a separate discussion.
This was a specific question, that already had a specific answer

If you feel the title of this discussion is too general, thats fair and we can discuss changing it

**Reply by chrisjsewell** [reply]
> I wanted to reference the fact that beyond the MEP process we don't have a cross-platform solution for changing syntax

and also I would suggest this is not quite correct; I created markdown-it-py with the exact purpose of having compatibility between python and js, and indeed all of the `markdown-it-xxx` package in this organisation derive from extensions in mdit-py-plugins

**Reply by agoose77** [reply]
> I'm happy to discuss user extensions more broadly, but then it would be best to open a separate discussion.
This was a specific question, that already had a specific answer

Personally I felt that the issue was general enough that it warranted further context. I agree that any additional discussion might be better suited in a dedicated thread, but this issue will still appear in the search given the content, so that's where I pitched my reply. It actually took me a second pass to realise that there was a specific reference to `myst-parser` configuration that scoped it to that tool!

> and also I would suggest this is not quite correct; I created markdown-it-py with the exact purpose of having compatibility between python and js, and indeed all of the markdown-it-xxx package in this organisation derive from extensions in mdit-py-plugins

I think this is semantics. A third-party user can't currently write a single blob of code that extends the MyST syntax across all frontends, and that's what I'm referring to here. The fact that we have a Python port of markdown-it is fantastic, because it means adding new syntax is a lot easier.

**Reply by chrisjsewell** [reply]
> A third-party user can't currently write a single blob of code that extends the MyST syntax across all frontends, and that's what I'm referring to here. 

by front ends, do you mean HTML/LaTeX/JATS, etc?
The way to do that is to have a defined set of node types, which is what pandoc and sphinx/docutils have, although sphinx also allows you to introduce new nodes, which you then have to specify builder specific rendering logic (or have fallbacks)

In terms of source to AST parsing; if the spec is clear enough anyone can implement it
As I've mentioned before, djot has at least five different implementations, over four different programming languages. They've all managed to implement the spec, and change with it as its evolved

**Reply by agoose77** [reply]
> by front ends, do you mean HTML/LaTeX/JATS, etc?

Ugh, you go to one Jupyter workshop and "frontend" becomes every third word :laughing: 

> The way to do that is to have a defined set of node types, which is what pandoc and sphinx/docutils have, although sphinx also allows you to introduce new nodes, which you then have to specify builder specific rendering logic (or have fallbacks)

Yes, and that's what I was touching upon with the idea that arbitrary roles and directives are soft extension targets that we can more easily support across renderers, as they have graceful (enough) fallbacks and any extension mechanism can be designed such that it maps roles/directives to a custom renderer. Whereas, custom parsing is a whole other kettle of fish.

> In terms of source to AST parsing; if the spec is clear enough anyone can implement it

That's my hope for the spec — parsers can come and go as long as they produce our AST and don't hurt anyone! :P

**Reply by tavin** [reply]
Well even being able to add just roles *and* directives in a pluggable way would be helpful. Indeed it seems clear enough that rendering can fall back to code blocks or divs with a class.

As the scope of this discussion has grown, let me add a couple remarks.

Personally I don't care if I have to reimplement my custom code across different parse/render toolchains. That's normal.

I also don't care if I have written an extension based on a specific platform (e.g. the markdownit parser) and then upstream changes to a new platform. In my view, as a custom extension author it's my task to adapt my code, and there is no interface contract.

So whether we're talking about roles/directives or my original ask, I'd rather have something limited by these 2 caveats than not have it at all.

**Reply by tavin** [reply]
See also https://github.com/executablebooks/jupyterlab-myst/pull/115 which opens a pathway to custom directives. It would become possible to write a separate labextension adding directive specs when the myst parser is called.
